docs/datatypes/tpromise.md

```diff
@@ -0,0 +1,90 @@
+---
+id: datatypes_tpromise
+title: "TPromise"
+---
+
+`TPromise` is a mutable reference that can be set exactly once and can participate in transactions in STM.
+
+## Create a TPromise
+
+Creating a `TPromise`:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tPromise: STM[Nothing, TPromise[String, Int]] = TPromise.make[String, Int]
+```
+
+## Complete a TPromise
+
+In order to successfully complete a `TPromise`:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tPromiseSucceed: UIO[TPromise[String, Int]] = for {
+  tPromise <- TPromise.make[String, Int].commit
+  _        <- tPromise.succeed(0).commit
+} yield tPromise
+```
+
+In order to fail a `TPromise` use:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tPromiseFail: UIO[TPromise[String, Int]] = for {
+  tPromise <- TPromise.make[String, Int].commit
+  _        <- tPromise.fail("failed").commit
+} yield tPromise
+```
+
+Alternatively, you can use `Either[E, A]` in combining with a `done` function:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tPromiseDoneSucceed: UIO[TPromise[String, Int]] = for {
+  tPromise <- TPromise.make[String, Int].commit
+  _        <- tPromise.done(Right(0)).commit
+} yield tPromise
+
+val tPromiseDoneFail: UIO[TPromise[String, Int]] = for {
+  tPromise <- TPromise.make[String, Int].commit
+  _        <- tPromise.done(Left("failed")).commit
+} yield tPromise
+```
+
+Note that a value can be set only once and every following attempt will return `false` as a result.
```

Comment:
Replace with "Once the value is set, following attempts to set it will result in `false`".